### PR TITLE
Implement ZIP 215 validation rules.

### LINKED ISSUE #16: Implement ZIP 215.
[ZIP 215](https://github.com/zcash/zips/blob/master/zip-0215.rst) will change the validation rules for Ed25519 signatures on Zcash.  As this library aims to provide Zcash-flavored Ed25519, it necessarily must also be updated in sync with the ZIP 215 activation.

I would very much prefer not to maintain multiple code paths, so the current plan is:

* release a version `1.x` series with the current validation rules;
* release a version `2.x` series with the ZIP 215 validation rules.

Since downstream crates can import multiple versions of the same dependency simultaneously (by renaming the dependency so that the two versions have different names), it's still possible to use different validation rules in different contexts if someone needed to do that (e.g., pre/post activation), but the API remains simple for everyone who doesn't particularly care about the details of the validation rules.

### PR DESCRIPTION
**Pull request by hdevalence**
Closes #16.

Done:
- [x] validation criteria
- [x] testing
  - [x] test vectors that exercise the differences between cofactorless and cofactored verification equations;
  - [x] test vectors that exercise non-canonical point encodings.

### CODE REVIEW

**Comment by hdevalence** [comment]
A hex-formatted list of test vectors can be found here: 
https://gist.github.com/hdevalence/93ed42d17ecab8e42138b213812c8cc7

This should make its way into ZIP 215.

File: src/verification_key.rs [hunk 1]
```diff
@@ -68,21 +69,11 @@ impl From<VerificationKeyBytes> for [u8; 32] {
...
-///
-/// but `libsodium` `1.0.15` does not check this. Instead it only checks whether
-/// the encoding of `A` is an encoding of a point on the curve and that the
-/// encoding is not all zeros. This implementation matches the `libsodium`
-/// behavior.  This has implications for signature verification behaviour, as noted
-/// in the [`VerificationKey::verify`] documentation.
+/// * `A_bytes` MUST be an encoding of a point `A` on the twisted Edwards form of
+///   Curve25519, and non-canonical encodings MUST be accepted;
 ///
 /// [ps]: https://zips.z.cash/protocol/protocol.pdf#concretejssig
```
Comment (ebfull) [inline src/verification_key.rs:78]: This moved to https://zips.z.cash/protocol/protocol.pdf#concreteed25519

File: src/verification_key.rs [hunk 2]
```diff
@@ -148,44 +134,32 @@ impl VerificationKey {
...
+    /// * the verification equation `[8][s]B = [8]R + [8][k]A` MUST be satisfied;
     ///
-    /// * The public key bytes must not be all 0, per `libsodium` `1.0.15`
-    /// `crypto_sign/ed25519/ref10/open.c:138-143`, which we maintain as an
-    /// invariant on the `VerificationKey` type.
+    /// * the alternate verification equation `[s]B = R + [k]A`, allowed by RFC 8032, MUST NOT be
+    ///   used.
     ///
-    /// [ps]: https://zips.z.cash/protocol/protocol.pdf#concretejssig
+    /// [ps]: https://zips.z.cash/protocol/protocol.pdf#concretejssig [ZIP 215]:
```
Comment (ebfull) [inline src/verification_key.rs:152]: This moved to https://zips.z.cash/protocol/protocol.pdf#concreteed25519

File: src/verification_key.rs [hunk 3]
```diff
@@ -194,10 +168,13 @@ impl VerificationKey {
...
-        // We expect to recompute R as [s]B - [k]A = [k](-A) + [s]B.
-        let R = EdwardsPoint::vartime_double_scalar_mul_basepoint(&k, &self.minus_A, &s);
+        //       [8][s]B = [8]R + [8][k]A
+        // <=>   [8]R = [8][s]B - [8][k]A
+        // <=>   0 = [8](R - ([s]B - [k]A))
+        // <=>   0 = [8](R - R')  where R' = [s]B - [k]A
+        let R_prime = EdwardsPoint::vartime_double_scalar_mul_basepoint(&k, &self.minus_A, &s);
 
-        if R.compress().to_bytes() == signature.R_bytes {
+        if (R - R_prime).mul_by_cofactor().is_identity() {
```
Comment (ebfull) [inline src/verification_key.rs:177]: Why not do `is_small_order()`?

Comment (hdevalence) [reply src/verification_key.rs:177]: No particular reason, I think I was trying to match the description in the spec more closely.

File: src/verification_key.rs (hunk 1 above)
Comment (daira) [reply src/verification_key.rs]: The old link is still valid (I try to keep links valid unless there's been a really big refactoring), but yeah.

Comment (hdevalence) [reply src/verification_key.rs]: I've updated the links anyways to point to the new version.

File: src/verification_key.rs (hunk 2 above)
Comment (hdevalence) [reply src/verification_key.rs]: :+1:

**Pull request merged**